### PR TITLE
Expose per-detection logits in `Results` with `logits=True`

### PR DESCRIPTION
**Pull request by Burhan-Q**
Proposing changes to expose `logits` output for standard model inference. Detection `logits` will be especially useful for out of distribution (OOD) label analysis, but could also be used for knowledge distillation, active learning, soft-label generation for semi-supervised detection, hierarchical and/or multi-label post-processing, etc.

Will not work on end-to-end models (YOLO26), semantic, pose, OBB, and/or exported models. Despite the limitations, the other tasks operate the same (no-op), and the benefits are still considerable for the use cases where `logits` is supported.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds an opt-in `logits=True` prediction flag to expose per-detection raw class scores in `Results`, with docs, config support, and tests across detection, segmentation, pose, and OBB workflows. 🚀

### 📊 Key Changes
- Added a new `logits` prediction argument to the config and prediction docs, documented as attaching per-detection pre-sigmoid class scores to `results[i].logits`.
- Extended `Results` with a new `logits` field and introduced a `Logits` container class with a `.probs` view for sigmoid-converted probabilities.
- Updated detection postprocessing to gather raw class scores aligned with surviving NMS detections when `logits=True` is enabled.
- Added segmentation handling to preserve raw scores before tuple unwrapping so the same logits flow works for segment models.
- Implemented graceful fallback with a warning when raw scores are unavailable, such as end-to-end models or exported models.
- Added Python tests covering supported model types (`yolo11n`, `yolo11n-seg`, `yolo11n-pose`, `yolo11n-obb`) and verifying logits are ignored safely for end-to-end models.

### 🎯 Purpose & Impact
- Enables advanced downstream use cases like confidence calibration, knowledge distillation, OOD detection, and active learning. 🎯
- Improves the inference API by exposing richer per-detection metadata without changing default behavior.
- Keeps compatibility strong by making the feature opt-in and automatically disabling it when model outputs do not provide raw class scores.
- Expands test coverage and documentation, reducing regression risk and making the new capability easier for users to adopt. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Burhan-Q, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below for a smooth integration. 😊

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The diff looks clean overall. The `logits` flag is wired through config, docs, result containers, and predictor postprocessing in a consistent way, and the unsupported-model fallback is handled gracefully. I did not find any clear bugs or actionable issues in the changed lines.

**Comment by Burhan-Q** [comment]
Related issues:
* https://github.com/ultralytics/ultralytics/issues/2423
* https://github.com/ultralytics/ultralytics/issues/20927
* https://github.com/ultralytics/ultralytics/issues/4574
* https://github.com/ultralytics/ultralytics/issues/2600

Possibly related issues:
* https://github.com/ultralytics/ultralytics/issues/4353
* https://github.com/ultralytics/ultralytics/issues/6197
* http://github.com/ultralytics/ultralytics/issues/17013
* https://github.com/ultralytics/ultralytics/issues/8995
* https://github.com/ultralytics/ultralytics/issues/15679
* https://github.com/ultralytics/ultralytics/issues/17101

**Comment by Burhan-Q** [comment]
FYI: @fcakyon & @Laughing-q

**Comment by Burhan-Q** [comment]
image
Latest CI failure unrelated to changes made by this PR, as previous merges from main did not cause CI failures. Other PRs are showing the same failures:

* [23105](https://github.com/ultralytics/ultralytics/actions/runs/26823712110/job/79120445726)
* [23101](https://github.com/ultralytics/ultralytics/actions/runs/26817959951/job/79064512015)
* [23103](https://github.com/ultralytics/ultralytics/actions/runs/26819111900/job/79068560934)

File: ultralytics/models/yolo/segment/predict.py [hunk 1]
```diff
@@ -61,6 +61,10 @@ def postprocess(self, preds, img, orig_imgs):
             >>> predictor = SegmentationPredictor(overrides=dict(model="yolo26n-seg.pt"))
             >>> results = predictor.postprocess(preds, img, orig_img)
         """
+        # Stash raw class scores for logits=True before the segment-specific tuple is unwrapped.
+        if getattr(self.args, "logits", False):
+            self._raw_scores = self._extract_raw_scores(preds)
```
Comment (fcakyon) [inline ultralytics/models/yolo/segment/predict.py]: this stashes scores in nms order but construct_result below drops empty-mask rows via the keep filter (around line 113) without filtering logits, so r.logits gets longer than r.boxes whenever a mask is empty (repro yolo11n-seg imgsz=480 conf=1e-6 gives boxes=299 vs logits=300)

minimal fix is to skip that keep filter when logits is on, same as the _feats branch already does

```python
if not (all(keep) or getattr(self, "_feats", None) is not None or getattr(self.args, "logits", False)):
```

**Review by fcakyon — COMMENTED** [review]
thanks for the pr @Burhan-Q, i think its a good idea. i have a few suggestions

**Comment by glenn-jocher** [comment]
Thanks @Burhan-Q for the contribution, the linked use cases, and the work keeping this PR updated. Easier access to per-detection class logits is a useful request.

After reviewing the full current diff and running focused validation, we are closing this implementation because it does not meet our current correctness and minimal-complexity bar.

There is a reproducible alignment bug: `yolo11n-seg.pt` on the bundled `bus.jpg` with `imgsz=320, conf=0.001, logits=True` returns 149 boxes but 150 logits rows. The segment predictor filters empty masks during result construction, after the logits have been selected, so the promised correspondence between logits and surviving detections is lost.

The existing raw model outputs and NMS retained indices already support extracting aligned detection logits; we verified that path on YOLO11 and YOLO26. Adding a global argument, a public container, and predictor state needs a stronger justification for the permanent API and maintenance cost, particularly while NMS-free and exported inference remain unsupported.

We remain open to a smaller design that uses the existing owners and preserves alignment through final result filtering. Thank you again for bringing the use cases together and contributing your time.
